### DISCUSSION TITLE
share global state between extensions

### DISCUSSION THREAD

**Discussion by chnliquan** [eclipse-theia/theia#10315] [Q&A]
how can i share global state bewteen extensions，eg userinfo and shared data generated dymaticilly by any extension

**Comment by maheshrajrp** [comment]
I'm not the expert here. But, it is possible to share Global State between extension using the backend.

**Reply by chnliquan** [reply]
i want to share global state in brower runtime.its deference from backend data

**Comment by chnliquan** [comment]
@msujew @vince-fugnitto can you help me🙏

**Reply by msujew** [reply]
Is this about Theia-Extensions (i.e. compile time components of Theia) or Theia-Plugins (vscode-extension)?

If you refer to Theia extensions, you can simply create a npm project in your Theia monorepo that's consumed by every one of your extensions. This project could contain some sort of static map/object to store data across extensions in the browser frontend.

Talking about vscode extensions, you can build a vscode extension that does the same (holding data in a map for example) and reference it by other vscode extensions using the [`vscode.extensions` API](https://code.visualstudio.com/api/references/vscode-api#extensions)

**Reply by chnliquan** [reply]
i want to share global state between all Theia-Extensions in my project, the shared state is  in browser frontend runtime, it want to implement some like react global state share by React.Context, but I didn’t find a suitable way in theia

**Reply by msujew** [reply]
> you can simply create a npm project in your Theia monorepo that's consumed by every one of your extensions. This project could contain some sort of static map/object to store data across extensions in the browser frontend.

That should be a suitable solution to your problem. Using the proposed way, you'll have a single API that you can use across your Theia extensions to store/retrieve in-memory data. In the end, as Theia extensions are compile time dependencies, you have complete control over any global/static fields across your application. How you deal with accessing/storing data (especially in memory) is completely your choice. Theia does not impose any restrictions on that. 

As a consequence, we cannot really propose a one-size-fits-all solution here, as it depends on your setup/architecture and personal preferences.

**Reply by chnliquan** [reply]
In this way, each extension has its own data, and then provides data communication through EventEmitter. If other extensions need the data provided by this extension, they need to monitor its changes. I don’t know if I understand it correctly.

 if i understand correctly, there is a problem between extensions No longer a black box, extension provider need to know exactly what data each extensions will generated

**Comment by chnliquan** [comment]
For example, user information needs to be shared throughout the theia extentions. Or the `A` extension has a state `a` that needs to be used by the `B`, `C`, and `D` extension, its similar to the relationship between `redux` and `react`

**Comment by chnliquan** [comment]
I really need your help 🙏 @msujew @colin-grant-work @vince-fugnitto

**Comment by colin-grant-work** [comment]
Another possibility is to modify application configuration to include the data you need. The application configuration is primarily derived from the `theia.frontend.config` object in your application's `package.json` but it's stored on the `window` object, so it can be accessed by anyone anywhere, and the same instance is passed to anyone who asks for it, so it can store global state if you'd like. 

https://github.com/eclipse-theia/theia/blob/master/packages/core/src/browser/frontend-application-config-provider.ts

**Reply by chnliquan** [reply]
There is no automatic notification mechanism in this way. I hope that an extension that relies on this part of the data after the global data changes can receive a reminder of his change. There is a solution that each extension maintains an EventEmitter, so there is a problem between extensions No longer a black box, you need to know exactly what data each extensions will generate @colin-grant-work
